community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexingService.java

```diff
@@ -437,7 +437,7 @@ private void apply( Iterable<IndexEntryUpdate<LabelSchemaDescriptor>> updates, I
      * it is *vital* that it is stable, and handles errors very well. Failing here means that the entire db
      * will shut down.
      */
-    public void createIndexes( IndexRule... rules ) throws IOException
+    public synchronized void createIndexes( IndexRule... rules ) throws IOException
```

Comment:
Only the `createIndexes` and the `dropIndex` synchronise on the `IndexingService` monitor lock as far as I can see. This means the state in here is inconsistently synchronised. If we are only worried about concurrent updates to the `indexMapRef`, then we should synchronise on that, and do so consistently. This synchronisation could, for instance, move into the `IndexMapReference` class and be exposed as an atomic `update` method that takes a lambda, or something like that.

Reply:
Maybe. Which are the other usages that should have been synchronized?

Reply:
It's not that `init` and `start` _need_ to synchronise. Rather, I'd like the synchronisation to be consistent, e.g. by pushing the responsibility down to the `IndexMapReference`, such that there is no way to get it wrong.

Reply:
Right OK, that's of course never a bad idea. I can have a look at that suggestion about a lambda, I like it.